Add a pressure drop postprocessor to NS module

GiudGiud opened the pull request:
and also use it to control a diode dynamically and show how it can be done

refs https://github.com/idaholab/moose/issues/23685

GiudGiud committed:
Add a new example for the friction diode including several Control strategies, refs #20696

GiudGiud committed:
Add a pressure drop postprocessor, refs #23685

GiudGiud committed:
Add docs for the pressure drop PP

GiudGiud committed:
Add more info to diode test

GiudGiud committed:
Detect if face is upstream or downstream before contributing to the pressure integrals

GiudGiud committed:
Handle FV interpolation method in a shared routine across pp and advection kernels, refs #23685

GiudGiud committed:
Make weighting functor optional, refs #23685

GiudGiud committed:
Add test for pressure drop postprocessor, refs #23685

GiudGiud committed:
Add test for weighted pressure drop with finite volume, refs #23685

GiudGiud committed:
Add test for weighted pressure drop with finite element NS, refs #23685

GiudGiud committed:
Add documentation on control of the diode operation, refs #23685

GiudGiud committed:
Add diode test for three control strategies refs #20696

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1405168/) on d49132e wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/23726/site)

This comment will be updated on new commits.

GiudGiud committed:
Fix test failures:
- use ad_indexing global for FE functors
- allow warnings on a distributed use of a MG
Fix documentation links

lindsayad commented on modules/navier_stokes/src/utils/NSFVUtils.C:
```diff
@@ -22,31 +22,8 @@ setInterpolationMethods(const MooseObject & obj,
-
-    need_more_ghosting = true;
-  }
+  bool need_more_ghosting =
```
```suggestion
  const bool need_more_ghosting =
```

lindsayad commented on modules/navier_stokes/src/utils/NSFVUtils.C:
```diff
@@ -22,31 +22,8 @@ setInterpolationMethods(const MooseObject & obj,
-    need_more_ghosting = true;
-  }
+  bool need_more_ghosting =
+      setInterpolationMethod(obj, advected_interp_method, "advected_interp_method");
```
Nice, I like this change

lindsayad commented on modules/navier_stokes/test/tests/finite_volume/materials/flow_diode/tests:
```diff
@@ -1,13 +1,48 @@
+      csvdiff = 'time_based.csv'
+      cli_args = 'Controls/active=time_based Outputs/file_base=time_based'
+      ad_indexing_type = 'global'
+      detail = 'with a simple time-based turn-on criterion'
```
```suggestion
      detail = 'with a simple time-based turn-on criterion,'
```
@cticenhour explained to me this is the proper way to do details

lindsayad commented on modules/navier_stokes/test/tests/finite_volume/materials/flow_diode/tests:
```diff
@@ -1,13 +1,48 @@
+      csvdiff = 'pdrop_based.csv'
+      cli_args = 'Controls/active=pdrop_based Outputs/file_base=pdrop_based'
+      ad_indexing_type = 'global'
+      detail = 'with a simple pressure-based turn-on criterion'
```
```suggestion
      detail = 'with a simple pressure-based turn-on criterion,'
```

lindsayad commented on modules/navier_stokes/test/tests/finite_volume/materials/flow_diode/tests:
```diff
@@ -1,13 +1,48 @@
+      csvdiff = 'flow_based.csv'
+      cli_args = 'Controls/active=flow_based Outputs/file_base=flow_based'
+      ad_indexing_type = 'global'
+      detail = 'with a simple flow-based turn-on criterion'
```
```suggestion
      detail = 'with a simple flow-based turn-on criterion.'
```

lindsayad commented on modules/navier_stokes/test/tests/finite_volume/materials/flow_diode/transient_operation.i:
```diff
@@ -0,0 +1,307 @@
+
+    friction_blocks = 'fluid; diode'
+    friction_types = 'darcy forchheimer; darcy forchheimer'
+    # Base friction
+    # friction_coeffs = 'Darcy Forchheimer; Darcy Forchheimer'
```
You want these in?

GiudGiud replied:
yep it shows the difference

lindsayad commented on modules/navier_stokes/test/tests/finite_volume/materials/flow_diode/transient_operation.i:
```diff
@@ -0,0 +1,307 @@
+    friction_coeffs = 'combined_linear combined_quadratic; combined_linear combined_quadratic'
+
+    # Porosity jump treatment
+    # Option 1: diffusion correction
+    use_friction_correction = true
+    consistent_scaling = 10
+
+    # Option 2: bernouilli jump
+    # porosity_interface_pressure_treatment = bernoulli
```
Do you want to mention porosity smoothing if you're going to list options?

GiudGiud replied:
I kind of mean it as advertisement, and if we want to try something else here

lindsayad commented on modules/navier_stokes/test/tests/finite_volume/materials/flow_diode/transient_operation.i:
This is a really nice demonstrative and explanatory input; thank you!

GiudGiud replied:
Thanks! Wish I had done this earlier

lindsayad commented on modules/navier_stokes/test/tests/postprocessors/pressure_drop/tests:
```diff
@@ -0,0 +1,134 @@
+      input = drop_insad.i
+      csvdiff =  diverging_insad.csv
+      abs_zero = 1e-8
+      detail = 'with a regular face pressure evaluation'
```
```suggestion
      detail = 'with a regular face pressure evaluation,'
```

lindsayad commented on modules/navier_stokes/test/tests/postprocessors/pressure_drop/tests:
```diff
@@ -0,0 +1,134 @@
+      cli_args = "Outputs/file_base='diverging_insad_weighted'
+                  Postprocessors/pdrop_total/weighting_functor='velocity' Postprocessors/pdrop_mid1/weighting_functor='velocity'
+                  Postprocessors/pdrop_mid2/weighting_functor='velocity' Postprocessors/pdrop_mid3/weighting_functor='velocity'"
+      detail = 'with a pressure drop face evaluation weighted by the local velocity'
```
```suggestion
      detail = 'with a pressure drop face evaluation weighted by the local velocity.'
```

lindsayad commented on modules/navier_stokes/test/tests/postprocessors/pressure_drop/tests:
```diff
@@ -0,0 +1,134 @@
+  []
+
+  [insfv_straight]
+    requirement = 'The system shall be able to compute the pressure drop in a frictionless heated straight channel with a finite volume incompressible Navier Stokes model.'
```
```suggestion
    requirement = 'The system shall be able to compute the pressure drop in a frictionless heated straight channel with a finite volume incompressible Navier Stokes model'
```

lindsayad commented on modules/navier_stokes/test/tests/postprocessors/pressure_drop/tests:
```diff
@@ -0,0 +1,134 @@
+      csvdiff =  straight_insfv.csv
+      abs_zero = 1e-8
+      cli_args = "Mesh/inactive='diverging_mesh' Outputs/file_base='straight_insfv' -pc_type lu -pc_factor_shift_type NONZERO"
+      detail = 'with a regular face pressure evaluation'
```
```suggestion
      detail = 'with a regular face pressure evaluation,'
```

lindsayad commented on modules/navier_stokes/test/tests/postprocessors/pressure_drop/tests:
```diff
@@ -0,0 +1,134 @@
+                  -pc_type lu -pc_factor_shift_type NONZERO
+                  Postprocessors/pdrop_total/weighting_functor='velocity' Postprocessors/pdrop_mid1/weighting_functor='velocity'
+                  Postprocessors/pdrop_mid2/weighting_functor='velocity' Postprocessors/pdrop_mid3/weighting_functor='velocity'"
+      detail = 'with a pressure drop face evaluation weighted by the local velocity'
```
```suggestion
      detail = 'with a pressure drop face evaluation weighted by the local velocity.'
```

lindsayad replied:
@cticenhour should there be a conjunction on the last detail? (e.g. and)

cticenhour replied:
APA style applies `, and` at the end of the next-to-last detail to make the itemized list a complete sentence. I have been following that approach.

lindsayad commented on modules/navier_stokes/test/tests/postprocessors/pressure_drop/tests:
```diff
@@ -0,0 +1,134 @@
+      input = drop_insfv.i
+      csvdiff = diverging_insfv_quad_rz.csv
+      abs_zero = 1e-8
+      detail = 'with a quadrilateral mesh in RZ geometry,'
```
```suggestion
      detail = 'with a quadrilateral mesh in RZ geometry.'
```

lindsayad commented on modules/navier_stokes/test/tests/postprocessors/pressure_drop/tests:
```diff
@@ -0,0 +1,134 @@
+      input = 'drop_insad.i'
+      expect_err = 'Upstream boundary \'bottom\' is not included in boundary restriction'
+      cli_args = "Postprocessors/pdrop_total/boundary='top'"
+      detail = 'an upstream boundary for the pressure is not a boundary for the postprocessor'
```
```suggestion
      detail = 'an upstream boundary for the pressure is not a boundary for the postprocessor,'
```

lindsayad commented on modules/navier_stokes/test/tests/postprocessors/pressure_drop/tests:
```diff
@@ -0,0 +1,134 @@
+      input = 'drop_insad.i'
+      expect_err = 'Boundary restriction on boundary \'internal_top\' is not part of upstream or downstream boundaries'
+      cli_args = "Postprocessors/pdrop_total/boundary='top bottom internal_top'"
+      detail = 'a boundary for the postprocessor is not part of either the upstream or downstream pressure evaluation'
```
```suggestion
      detail = 'a boundary for the postprocessor is not part of either the upstream or downstream pressure evaluation,'
```

lindsayad commented on modules/navier_stokes/test/tests/postprocessors/pressure_drop/tests:
```diff
@@ -0,0 +1,134 @@
+      input = 'drop_insad.i'
+      expect_err = 'Upstream boundary \'top\' is also a downstream boundary'
+      cli_args = "Postprocessors/pdrop_total/upstream_boundary='top bottom'"
+      detail = 'a downstream boundary is also an upstream boundary for the pressure drop'
```
```suggestion
      detail = 'a downstream boundary is also an upstream boundary for the pressure drop,'
```

lindsayad commented on modules/navier_stokes/test/tests/postprocessors/pressure_drop/tests:
```diff
@@ -0,0 +1,134 @@
+                  Materials/null/prop_names='null'
+                  Materials/null/prop_values='0 0 0'
+                  Postprocessors/pdrop_total/weighting_functor=null"
+      detail = 'the weighting functor integrates to 0'
```
```suggestion
      detail = 'the weighting functor integrates to 0,'
```

lindsayad commented on modules/navier_stokes/test/tests/postprocessors/pressure_drop/tests:
```diff
@@ -0,0 +1,134 @@
+      input = 'drop_insad.i'
+      expect_err = 'Face interpolation only specified for finite volume'
+      cli_args = "Postprocessors/pdrop_total/weighting_interp_method='average'"
+      detail = 'a face interpolation rule is specified for a finite element pressure variable'
```
```suggestion
      detail = 'a face interpolation rule is specified for a finite element pressure variable.'
```

lindsayad commented on modules/navier_stokes/test/tests/postprocessors/pressure_drop/tests:
very impressive error checking!

lindsayad commented on modules/navier_stokes/test/tests/postprocessors/pressure_drop/tests:
```diff
@@ -0,0 +1,134 @@
+      input = drop_insfv.i
+      csvdiff = diverging_insfv_quad_xy.csv
+      abs_zero = 1e-8
+      detail = 'with a quadrilateral mesh in XY geometry, with mass flow measured using either a variable or material property,'
```
```suggestion
      detail = 'with a quadrilateral mesh in XY geometry, with mass flow measured using either a variable or material property, and'
```

GiudGiud committed:
Apply suggestions from code review

- fix up detail & requirements
- improved logic in if-else
- improved constness
- add a routine for all interpolation methods in NS
- avoid Qp evaluations in PressureDrop PP by not relying on Coupleable interface
- make doco cleaer
Co-authored-by: Alex Lindsay <alexlindsay239@gmail.com>

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1410274/) on 81a04b0 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/23726/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/23726/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format ae61f4e8046d6aa06bc153d0cfcf0c729c297195`

GiudGiud commented on modules/navier_stokes/doc/content/source/postprocessors/PressureDrop.md:
```diff
@@ -0,0 +1,43 @@
+then the pressure drop is positive.
+
+!alert note
+If using mesh refinement, upstream or downstream boundaries may not be split between two refined elements,
+children of the same parent element.
```
reworded. I dont expect this to ever happen

GiudGiud commented on modules/navier_stokes/doc/content/source/materials/NSFVFrictionFlowDiodeMaterial.md:
```diff
@@ -17,6 +23,30 @@ The friction coefficients are then combined using a [PiecewiseByBlockVectorFunct
+!listing test/tests/finite_volume/materials/flow_diode/transient_operation.i block=Controls/time_based Functions/time_function
+
+The second strategy is to look at the pressure drop across the diode, and block (add friction to) the flow if it exceeds a certain value.
+If it exceeds a certain value, then it means the flow is flowing through the diode in a certain direction.
```
here it s as opposed to the other direction
reworded

GiudGiud commented on modules/navier_stokes/doc/content/source/postprocessors/PressureDrop.md:
```diff
@@ -0,0 +1,43 @@
+This is effectively a weighting flux, which should usually be set to a momentum functor so
+that the effective weighting factor is the local mass flux.
+
+\begin{equation}
+\begin{split}
+P_{upstream} &= \dfrac{\int_{\delta \Omega}} P(x,y,z) \vec{W(x,y,z) \cdot \vec{n}}{\dfrac{\int_{\delta \Omega}} \vec{W(x,y,z) \cdot \vec{n}}} \\
+P_{downstream} &= ... \\
+P_{drop} &= P_{upstream} - P_{downstream}
+\end{split}
+\end{equation}
```
fixed, maybe

GiudGiud commented on modules/navier_stokes/src/postprocessors/PressureDrop.C:
```diff
@@ -0,0 +1,350 @@
+      "downstream_boundary",
+      "The downstream surface (must also be specified in 'boundary' parameter");
+
+  MooseEnum interp_method("average upwind sou min_mod vanLeer quick skewness-corrected", "upwind");
```
will do

GiudGiud replied:
done

GiudGiud commented on modules/navier_stokes/src/postprocessors/PressureDrop.C:
```diff
@@ -0,0 +1,350 @@
+#else
+        if (std::find(_downstream_boundaries.begin(), _downstream_boundaries.end(), bdy) !=
+            _downstream_boundaries.end())
+          mooseError("Boundary '", _mesh.getBoundaryName(bdy), "' is both upstream and downstream");
+#endif
+      }
+#ifndef NDEBUG
+      if (upstream &&
+          std::find(_downstream_boundaries.begin(), _downstream_boundaries.end(), bdy) !=
+              _downstream_boundaries.end())
+        mooseError("Face info is part of both upstream and downstream boundaries");
```
this is just to be extra sure

I think you could fool the check in the constructor, by having boundaries overlap

lindsayad replied:
If you think the extra surety is worth the ugliness of the ifdefs to developers...

lindsayad replied:
they definitely make the code harder to understand upon quick examination

lindsayad replied:
hahaha I'm not sure I agree

lindsayad replied:
If it's unclear whether a conversation I started is resolved, let me make the determination in the future please

GiudGiud replied:
I think the error is great. It s inspired by defense in depth, I want to keep it. 

Sorry for resolving I usually only resolve trivial stuff

GiudGiud commented on modules/navier_stokes/src/postprocessors/PressureDrop.C:
```diff
@@ -0,0 +1,350 @@
+      "Computes the pressure drop between an upstream and a downstream boundary.");
+
+  // Using the Coupleable interface only to know whether FE or FV
+  params.addRequiredCoupledVar("pressure", "The pressure functor");
```
I think I can remove this using the subproblem. lemme see

GiudGiud replied:
should be fixed

GiudGiud commented:
Should be good now. Maybe just that equation on the website if I messed up

lindsayad commented on modules/navier_stokes/src/postprocessors/PressureDrop.C:
```diff
@@ -0,0 +1,339 @@
+Real
+PressureDrop::computeFaceInfoWeightedPressureIntegral(const FaceInfo * const fi) const
+{
+#ifdef MOOSE_GLOBAL_AD_INDEXING
```
Will have to monitor #23665 to tell whether to keep this

GiudGiud replied:
ok will remove

lindsayad commented on modules/navier_stokes/src/postprocessors/PressureDrop.C:
```diff
@@ -0,0 +1,339 @@
+    if (!fi->neighborPtr() || !_pressure.hasBlocks(fi->neighborPtr()->subdomain_id()))
+    {
+      const auto ssf = Moose::FaceArg(
+          {fi, limiterType(_weight_interp_method), true, correct_skewness, _current_elem});
```
Are you indicating sidedness here on purpose, e.g. if we are on an internal face, then you want to ensure you don't use data from the other side? That would be the appropriate use of the sided-element argument. But then if you are going to do that ... are there ever going to be times where `!_pressure.hasBlocks(fi->elemPtr()->subdomain_id())` ?

lindsayad replied:
I'm mostly wondering why you've kept the `!fi->neighborPtr() || !_pressure.hasBlocks(fi->neighborPtr()->subdomain_id())` condition around here when you've made the `if (_weighting_functor)` branch nice and "clean" (by just always passing `nullptr` to the sided element argument)

GiudGiud replied:
I think I misunderstood something critical about the FaceArg until now. I m going to adapt the docstring for it

GiudGiud replied:
and you can tell me if the misunderstanding persists

lindsayad replied:
the addition to the doc-string looks good

lindsayad commented on modules/navier_stokes/src/postprocessors/PressureDrop.C:
```diff
@@ -0,0 +1,339 @@
+  const bool correct_skewness =
+      _weight_interp_method == Moose::FV::InterpMethod::SkewCorrectedAverage;
+
+  if (_weighting_functor)
+  {
+    // External faces for the weighting functor
+    if (!fi->neighborPtr() || !_weighting_functor->hasBlocks(fi->neighborPtr()->subdomain_id()))
+    {
+      const auto ssf = Moose::FaceArg(
+          {fi,
+           limiterType(_weight_interp_method),
+           MetaPhysicL::raw_value((*_weighting_functor)(elem_arg)) * fi->normal() > 0,
+           correct_skewness,
+           _current_elem});
+      return fi->normal() * MetaPhysicL::raw_value((*_weighting_functor)(ssf));
+    }
+    else
+    {
+      const auto ssf = Moose::FaceArg(
+          {fi,
+           Moose::FV::limiterType(_weight_interp_method),
+           MetaPhysicL::raw_value((*_weighting_functor)(elem_arg)) * fi->normal() > 0,
+           correct_skewness,
+           nullptr});
+      return fi->normal() * MetaPhysicL::raw_value((*_weighting_functor)(ssf));
+    }
+  }
```
This logic is not consistent with the logic in `computeFaceInfoWeightedPressureIntegral`.
```suggestion
  if (_weighting_functor)
  {
    const auto ssf = Moose::FaceArg(
        {fi,
         Moose::FV::limiterType(_weight_interp_method),
         MetaPhysicL::raw_value((*_weighting_functor)(elem_arg)) * fi->normal() > 0,
         correct_skewness,
         nullptr});
    return fi->normal() * MetaPhysicL::raw_value((*_weighting_functor)(ssf));
  }
```

GiudGiud replied:
yeah this is what I want in the end

lindsayad commented on modules/navier_stokes/src/postprocessors/PressureDrop.C:
```diff
@@ -0,0 +1,339 @@
+  auto elem_arg = std::make_tuple(_current_elem, _qp, _qrule);
+  mooseAssert(_qp == 0, "Only one quadrature point");
+
+  if (_weighting_functor)
+  {
+    const auto ssf =
+        Moose::FaceArg({fi,
+                        Moose::FV::limiterType(_weight_interp_method),
+                        MetaPhysicL::raw_value((*_weighting_functor)(elem_arg)) * fi->normal() > 0,
+                        correct_skewness,
+                        nullptr});
+    const auto face_weighting = MetaPhysicL::raw_value((*_weighting_functor)(ssf));
+    return fi->normal() * face_weighting * _pressure(ssf);
+  }
+  else
+  {
+    // External faces for the pressure
+    if (!fi->neighborPtr() || !_pressure.hasBlocks(fi->neighborPtr()->subdomain_id()))
+    {
+      const auto ssf = Moose::FaceArg(
+          {fi, limiterType(_weight_interp_method), true, correct_skewness, _current_elem});
+      return _pressure(ssf);
+    }
+    else
+    {
+      const auto ssf = Moose::FaceArg(
+          {fi, Moose::FV::limiterType(_weight_interp_method), true, correct_skewness, nullptr});
+      return _pressure(ssf);
+    }
+  }
```
```suggestion
  const auto ssf =
      Moose::FaceArg({fi,
                      Moose::FV::limiterType(_weight_interp_method),
                      MetaPhysicL::raw_value((*_weighting_functor)(elem_arg)) * fi->normal() > 0,
                      correct_skewness,
                      nullptr});
  if (_weighting_functor)
  {
    const auto face_weighting = MetaPhysicL::raw_value((*_weighting_functor)(ssf));
    return fi->normal() * face_weighting * _pressure(ssf);
  }
  else
    return _pressure(ssf);
```

GiudGiud replied:
doesnt work for t he second case, cant use weighting functor to determine upstream-ness if not defined

GiudGiud committed:
Make slow new test not run with valgrind
and not run in 'dbg'

lindsayad commented:
Do you want me to run the thread failures through thread sanitizer?

GiudGiud commented:
no I fixed them already

GiudGiud commented:
I ll push the commit soon

GiudGiud commented:
thanks for offering though!

GiudGiud committed:
Address Alex's review:
- use the functor logic for handling the boundary case instead of building it in the postprocessor
- add docstrings
- remove the local AD indexing accomodation

GiudGiud committed:
Extend Alex's review to FlowRate PP

GiudGiud commented:
wow my "fix" for threading killed performance in debug

GiudGiud commented:
let's see if we can do that instead. Pretty sure the debug tests timing out were not threaded

lindsayad commented:
> let's see if we can do that instead

do what instead?

GiudGiud commented:
only lock if there's threads

GiudGiud commented on framework/src/mesh/MooseMesh.C:
```diff
@@ -3313,7 +3313,14 @@ MooseMesh::buildFiniteVolumeInfo() const
+  // but no need to lock if we are not in a threaded region
+  std::unique_ptr<Threads::spin_mutex::scoped_lock> lock;
+  if (Threads::in_threads)
+    lock = std::make_unique<Threads::spin_mutex::scoped_lock>(Threads::spin_mtx);
```
is there a cleaner way to do this?

The lock needs to stay in scope and I dont want to be creating it if we re not threaded (which we usually are not, but in my use here we are) because it's too slow (fails debug recipe pretty hard)

lindsayad replied:
We just shouldn't be calling this method in a threaded region. That's the fix: ensuring that the face information is already built before calling e.g. `faceInfo()` in a threaded region. Then there will be no need for locks

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1418066/) on 9b75b1a wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>ba109d</th>
      <th colspan=4>#23726 9b75b1</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.87%</td>
      <td>84.87%</td>
      <td>-0.00%</td>
      <td>50.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>84193</td>
      <td>84194</td>
      <td>+1</td>
      <td>3</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15009</td>
      <td>15013</td>
      <td>+4</td>
      <td>3</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/23726/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/23726/coverage/framework)


## Modules coverage

### Navier stokes

<table>
  <thead>
    <tr>
      <th></th>
      <th>ba109d</th>
      <th colspan=4>#23726 9b75b1</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>83.66%</td>
      <td>83.81%</td>
      <td>+0.15%</td>
      <td>96.99%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>10524</td>
      <td>10660</td>
      <td>+136</td>
      <td>161</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>2055</td>
      <td>2059</td>
      <td>+4</td>
      <td>5</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/23726/coverage_diff/navier_stokes)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/23726/coverage/navier_stokes)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/23726/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/23726/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/23726/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/23726/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/23726/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/23726/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/23726/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/23726/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/23726/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/23726/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/23726/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/23726/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/23726/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/23726/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/23726/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/23726/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/23726/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/23726/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/23726/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/23726/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/23726/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/23726/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/23726/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/23726/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/23726/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/23726/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/23726/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/23726/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 50.00% is less than the suggested 90.0%



This comment will be updated on new commits.

GiudGiud commented:
The new pressure drop tests are too slow for dbg as well. The locking seems like a huge slow down in debug serial.
The new diode tests are quite slow in dbg, clocking in at 200s and I dont think they bring anything coverage wise for debug so I disabled them there too

GiudGiud commented:
I m ok with that but then we need to explicitly ban threading in this function.
We can't run into the same problem I ran into

lindsayad commented:
How are you going to ban it? And should we explicitly ban all functions that aren't supposed to be run threaded from being run threaded? That's a lot of functions

GiudGiud commented:
I ll add an issue

GiudGiud commented:
Improving the API is worthwhile, but just tagging routines we know are not thread-safe will help and we can swap out the safety mechanism later

GiudGiud committed:
Prebuild face info from the pressure drop postprocessor
- avoid threaded building of face infos
- add an assert to prevent that building

lindsayad commented on modules/navier_stokes/test/tests/finite_volume/materials/flow_diode/tests:
```diff
@@ -1,13 +1,54 @@
+      input = 'transient_operation.i'
+      csvdiff = 'time_based.csv'
+      cli_args = 'Controls/active=time_based Outputs/file_base=time_based'
+      ad_indexing_type = 'global'
```
```suggestion
```

lindsayad commented on modules/navier_stokes/test/tests/finite_volume/materials/flow_diode/tests:
```diff
@@ -1,13 +1,54 @@
+      input = 'transient_operation.i'
+      csvdiff = 'pdrop_based.csv'
+      cli_args = 'Controls/active=pdrop_based Outputs/file_base=pdrop_based'
+      ad_indexing_type = 'global'
```
```suggestion
```

lindsayad commented on modules/navier_stokes/test/tests/finite_volume/materials/flow_diode/tests:
```diff
@@ -1,13 +1,54 @@
+      input = 'transient_operation.i'
+      csvdiff = 'flow_based.csv'
+      cli_args = 'Controls/active=flow_based Outputs/file_base=flow_based'
+      ad_indexing_type = 'global'
```
```suggestion
```

lindsayad commented on modules/navier_stokes/test/tests/finite_volume/materials/flow_diode/tests:
```diff
@@ -1,13 +1,54 @@
+      ad_indexing_type = 'global'
+      detail = 'with a simple time-based turn-on criterion,'
+      # see #21185
+      allow_warnings = true
```
What's warning?

GiudGiud replied:
https://github.com/idaholab/moose/issues/21185

lindsayad replied:
If there's a warning, why is the issue marked as closed?

lindsayad replied:
It wasn't clear to me from any of the text in the issue nor in the pull request titles that referenced the issue that a warning was the goal or being added. If the pull request closing an issue (in this case titled "More misc fixups") doesn't make clear why it's closing the issue, then a little more text explaining why we regard the issue as closed would be appreciated.

GiudGiud replied:
The PR was about fixing the bad code that was in place of this warning.
I only have to reference issues in commits and PRs, not explain each commit in the text again. If the reviewer finds it unclear what each commit is doing, they are welcome to ask me to reword the message.

lindsayad replied:
I just found the relevant commit on the issue. Things are clear to me now I think. So are we just getting lucky in our CI that our distributed mesh runs don't distribute the sideset across different procs?

GiudGiud replied:
I think we are lucky that the MG actually works with distributed meshes, but I have no idea why.

My guess is that the MG gets the nodeset right, then this gets converted to a sideset automatically ??

I m not sure but someone flagged this MG as not working with distributed meshes and I didnt have the expertise to overrule their opinion.

lindsayad replied:
I see. Understood. So it sounds like this test will likely continue to work even if the sideset gets partitioned across process boundaries.

lindsayad commented on modules/navier_stokes/test/tests/postprocessors/pressure_drop/tests:
```diff
@@ -0,0 +1,134 @@
+    abs_zero = 1e-8
+    requirement = 'The system shall be able to compute the pressure drop in a straight channel with a finite element incompressible Navier Stokes model.'
+    cli_args = "Mesh/inactive='diverging_mesh' Outputs/file_base='straight_insad'"
+    ad_indexing_type = 'global'
```
```suggestion
```

lindsayad commented on modules/navier_stokes/test/tests/postprocessors/pressure_drop/tests:
```diff
@@ -0,0 +1,134 @@
+      abs_zero = 1e-8
+      detail = 'with a regular face pressure evaluation, and'
+      cli_args = "Outputs/file_base='diverging_insad'"
+      ad_indexing_type = 'global'
```
```suggestion
```

lindsayad commented on modules/navier_stokes/test/tests/postprocessors/pressure_drop/tests:
```diff
@@ -0,0 +1,134 @@
+                  Postprocessors/pdrop_total/weighting_functor='velocity' Postprocessors/pdrop_mid1/weighting_functor='velocity'
+                  Postprocessors/pdrop_mid2/weighting_functor='velocity' Postprocessors/pdrop_mid3/weighting_functor='velocity'"
+      detail = 'with a pressure drop face evaluation weighted by the local velocity.'
+      ad_indexing_type = 'global'
```
```suggestion
```

lindsayad commented on modules/navier_stokes/test/tests/postprocessors/pressure_drop/tests:
```diff
@@ -0,0 +1,134 @@
+                  Postprocessors/pdrop_total/weighting_functor='velocity' Postprocessors/pdrop_mid1/weighting_functor='velocity'
+                  Postprocessors/pdrop_mid2/weighting_functor='velocity' Postprocessors/pdrop_mid3/weighting_functor='velocity'"
+      detail = 'with a pressure drop face evaluation weighted by the local velocity'
+      ad_indexing_type = 'global'
```
```suggestion
```

lindsayad commented on modules/navier_stokes/test/tests/postprocessors/pressure_drop/tests:
```diff
@@ -0,0 +1,134 @@
+      csvdiff = diverging_insfv_quad_xy.csv
+      abs_zero = 1e-8
+      detail = 'with a quadrilateral mesh in XY geometry, with mass flow measured using either a variable or material property, and'
+      ad_indexing_type = 'global'
```
```suggestion
```

lindsayad commented on modules/navier_stokes/test/tests/postprocessors/pressure_drop/tests:
```diff
@@ -0,0 +1,134 @@
+      abs_zero = 1e-8
+      detail = 'with a quadrilateral mesh in RZ geometry.'
+      cli_args = "Mesh/coord_type=RZ Outputs/file_base='diverging_insfv_quad_rz'"
+      ad_indexing_type = 'global'
```
```suggestion
```

GiudGiud committed:
Apply suggestions from code review

- remove indexing type tests spec

Co-authored-by: Alex Lindsay <alexlindsay239@gmail.com>

GiudGiud commented:
Thanks for the in depth reviews. I learned a lot about two of our APIs and that will definitely help me advise others!

lindsayad commented:
No problem, thanks for the great work!
